File: leetcode/letter-combinations-of-a-phone-number.py
from typing import List


class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        table = {
            "2": ["a", "b", "c"],
            "3": ["d", "e", "f"],
            "4": ["g", "h", "i"],
            "5": ["j", "k", "l"],
            "6": ["m", "n", "o"],
            "7": ["p", "q", "r", "s"],
            "8": ["t", "u", "v"],
            "9": ["w", "x", "y", "z"],
        }

        res = []

        for d in digits:
            current = []

            for c in table[d]:
                if len(res) > 0:
                    for r in res:
                        current.append(r + c)
                else:
                    current.append(c)

            res = current

        return res

File: leetcode/letter-combinations-of-a-phone-number.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
 public:
    vector<string> letterCombinations(string digits)
    {
        const string map[] = {
            "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz",
        };
        auto size = digits.size();

        vector<string> res;
        string cur;

        cur.reserve(size);

        function<void(int, string)> backtrack = [&](int l, string cur) {
            if (l >= size) {
                if (cur.size() > 0) {
                    res.push_back(cur);
                }
                return;
            }

            for (auto c : map[digits[l] - '0']) {
                cur.push_back(c);
                backtrack(l + 1, cur);
                cur.pop_back();
            }
        };

        backtrack(0, cur);

        return res;
    }
};
